paimon-flink/paimon-flink-cdc/src/main/java/org/apache/paimon/flink/sink/cdc/UpdatedDataFieldsProcessFunction.java

```diff
@@ -43,19 +50,51 @@ public class UpdatedDataFieldsProcessFunction
 
     private final Identifier identifier;
 
+    private Set<FieldIdentifier> latestFields;
+
     public UpdatedDataFieldsProcessFunction(
             SchemaManager schemaManager, Identifier identifier, Catalog.Loader catalogLoader) {
         super(catalogLoader);
         this.schemaManager = schemaManager;
         this.identifier = identifier;
+        this.latestFields = new HashSet<>();
     }
 
     @Override
     public void processElement(
             List<DataField> updatedDataFields, Context context, Collector<Void> collector)
             throws Exception {
-        for (SchemaChange schemaChange : extractSchemaChanges(schemaManager, updatedDataFields)) {
+        List<DataField> actualUpdatedDataFields =
+                updatedDataFields.stream()
+                        .filter(
+                                dataField ->
+                                        !latestDataFieldContain(new FieldIdentifier(dataField)))
+                        .collect(Collectors.toList());
+        if (CollectionUtils.isEmpty(actualUpdatedDataFields)) {
+            return;
+        }
+        for (SchemaChange schemaChange :
+                extractSchemaChanges(schemaManager, actualUpdatedDataFields)) {
             applySchemaChange(schemaManager, schemaChange, identifier);
         }
+        /**
+         * Here, actualUpdatedDataFields cannot be used to update latestFields because there is a
+         * non-SchemaChange.AddColumn scenario. Otherwise, the previously existing fields cannot be
+         * modified again.
+         */
+        updateLatestFields();
```

Comment:
Can we just add updatedDataFields to latestFields?

Reply:
**No**, this is the logic in previous versions, but there are risks in actual testing.
**Reason**: FieldIdentifier only guarantees the uniqueness of <name, type, description>. If any of the attributes is adjusted repeatedly, it will lead to misjudgment.
**For example**: Field A's type is int, and then it is changed to string. If you want to adjust it to int again, it will not work,because latestFields has already saved an element like <A, int, description>.